File: apps/windows/TeslaSync.App.Core/Push/DeviceRegistrationClient.cs
using System.Globalization;
using System.Net;
using System.Text;
using System.Text.Json;
using TeslaSync.App.Core.Auth;
using TeslaSync.App.Core.Data.Net;
using GeneratedApi = TeslaSync.Windows.Generated.Api;

namespace TeslaSync.App.Core.Push;

/// <summary>
/// The single real <see cref="IDeviceRegistrationClient"/> (P2/W6-0002). It POSTs/DELETEs against
/// TeslaSync's additive <c>/api/v1/devices</c> endpoint (ADR-009) over the same configured
/// <see cref="HttpClient"/> pipeline the generated client uses — so registration carries the W4
/// bearer token and the W5 resilience handler — applying the <c>/api/v1</c> version segment exactly
/// once, exactly like <c>GeneratedApiClient</c>.
///
/// <para><b>Contract note (no silent drift):</b> the device-registration endpoint is an additive
/// ADR-009 contract that the OpenAPI source-of-truth (<c>api/openapi/teslasync.openapi.json</c>) does
/// not yet expose, so it is absent from the generated <see cref="GeneratedApi.ApiEndpoints"/> table
/// (the runbook defers it for ADR review). This client therefore targets the ADR-009 path
/// directly via <see cref="DevicesPath"/>; it is the one seam to migrate to a generated
/// <see cref="GeneratedApi.EndpointDescriptor"/> the moment the contract is emitted — no other code
/// hardcodes the path.</para>
/// </summary>
public sealed class DeviceRegistrationClient : IDeviceRegistrationClient
{
    /// <summary>The additive ADR-009 device-registration route (versioned; the version is added once).</summary>
    public const string DevicesPath = "/devices";

    private readonly HttpClient _http;
    private readonly ApiClientOptions _options;
    private readonly Action<string>? _diagnostics;

    /// <summary>Creates the client over the shared (auth + resilience) <see cref="HttpClient"/> pipeline.</summary>
    public DeviceRegistrationClient(HttpClient http, ApiClientOptions options, Action<string>? diagnostics = null)
    {
        ArgumentNullException.ThrowIfNull(http);
        ArgumentNullException.ThrowIfNull(options);
        _http = http;
        _options = options;
        _diagnostics = diagnostics;
    }

    /// <inheritdoc />
    public async Task<DeviceRegistrationResponse> RegisterAsync(
        DeviceRegistrationRequest request,
        CancellationToken cancellationToken = default)
    {
        ArgumentNullException.ThrowIfNull(request);

        var uri = BuildUri(DevicesPath);
        using var message = new HttpRequestMessage(HttpMethod.Post, uri)
        {
            Content = new StringContent(
                JsonSerializer.Serialize(request, _options.Json),
                Encoding.UTF8,
                "application/json"),
        };

        Emit($"devices → POST {DevicesPath}");
        using var response = await _http.SendAsync(message, HttpCompletionOption.ResponseHeadersRead, cancellationToken)
            .ConfigureAwait(false);

        if (!response.IsSuccessStatusCode)
        {
            await ThrowForErrorAsync("register", response, cancellationToken).ConfigureAwait(false);
        }

        await using var stream = await response.Content.ReadAsStreamAsync(cancellationToken).ConfigureAwait(false);
        try
        {
            var decoded = await JsonSerializer
                .DeserializeAsync<DeviceRegistrationResponse>(stream, _options.Json, cancellationToken)
                .ConfigureAwait(false);
            if (decoded is null || string.IsNullOrEmpty(decoded.RegistrationId))
            {
                throw new ApiException("Device registration returned no registration id.");
            }

            return decoded;
        }
        catch (JsonException ex)
        {
            throw new ApiException("The device-registration response could not be decoded.", innerException: ex);
        }
    }

    /// <inheritdoc />
    public async Task UnregisterAsync(string registrationId, CancellationToken cancellationToken = default)
    {
        ArgumentException.ThrowIfNullOrEmpty(registrationId);

        var uri = BuildUri(DevicesPath + "/" + Uri.EscapeDataString(registrationId));
        using var message = new HttpRequestMessage(HttpMethod.Delete, uri);

        Emit($"devices → DELETE {DevicesPath}/{{id}}");
        using var response = await _http.SendAsync(message, HttpCompletionOption.ResponseHeadersRead, cancellationToken)
            .ConfigureAwait(false);

        // A missing registration is a successful unregister (idempotent sign-out cleanup).
        if (response.StatusCode == HttpStatusCode.NotFound)
        {
            return;
        }

        if (!response.IsSuccessStatusCode)
        {
            await ThrowForErrorAsync("unregister", response, cancellationToken).ConfigureAwait(false);
        }
    }

    private Uri BuildUri(string route)
    {
        var path = _options.VersionBasePath.TrimEnd('/') + "/" + route.TrimStart('/');
        var basePath = _options.BaseAddress.AbsolutePath;
        var combined = string.IsNullOrEmpty(basePath) || basePath == "/"
            ? path
            : basePath.TrimEnd('/') + "/" + path.TrimStart('/');

        return new UriBuilder(_options.BaseAddress) { Path = combined }.Uri;
    }

    private async Task ThrowForErrorAsync(string action, HttpResponseMessage response, CancellationToken cancellationToken)
    {
        var status = (int)response.StatusCode;
        string? code = null;
        string? snippet = null;
        try
        {
            var body = await response.Content.ReadAsStringAsync(cancellationToken).ConfigureAwait(false);
            if (!string.IsNullOrWhiteSpace(body))
            {
                snippet = Sanitize(body.Length <= 256 ? body : body[..256]);
                code = JsonSerializer.Deserialize<GeneratedApi.Error>(body, _options.Json)?.Code;
            }
        }
        catch (JsonException)
        {
            // Non-JSON error body; the status line is enough for diagnostics.
        }

        Emit($"devices ✗ {action} status={status.ToString(CultureInfo.InvariantCulture)}");
        throw new ApiException(
            $"Device {action} failed with status {status.ToString(CultureInfo.InvariantCulture)}.",
            status,
            snippet,
            code);
    }

    private void Emit(string line) => _diagnostics?.Invoke(Sanitize(line));

    private static string Sanitize(string line) => PushRedaction.Redact(TokenRedaction.Redact(line));
}

File: apps/android/app/src/main/java/io/teslasync/android/push/DeviceRegistrationClient.kt
package io.teslasync.android.push

import io.teslasync.shared.core.net.ApiError
import io.teslasync.shared.core.net.ApiHttpClient
import io.teslasync.shared.core.net.HttpMethodKind
import io.teslasync.shared.core.net.safeRequest

/**
 * The port the registration service uses to register and unregister this device with TeslaSync's
 * additive `/api/v1/devices` endpoint (P3/A6, ADR-009). It is a focused seam so the registration logic
 * is testable with a fake. Every method returns a non-throwing [Result].
 */
interface DeviceRegistrationClient {
    /** Registers (upserts) this device's channel; returns the backend registration response. */
    suspend fun register(request: DeviceRegistrationRequest): Result<DeviceRegistrationResponse>

    /**
     * Unregisters the device session identified by [registrationId] so the backend stops fanning push
     * to this channel. A `404` is treated as already-removed (idempotent sign-out cleanup).
     */
    suspend fun unregister(registrationId: String): Result<Unit>
}

/**
 * The single real [DeviceRegistrationClient] (P3/A6). It goes through the shared resilient
 * [ApiHttpClient] — so registration carries the A4 bearer token and the shared resilience handler, and
 * the `/api/v1` version segment is applied exactly once.
 *
 * The device-registration endpoint is an additive ADR-009 contract that the OpenAPI source-of-truth
 * does not yet expose (the backend route is provisioned in P5/H5-0001), so this client targets the
 * ADR-009 path directly via [DEVICES_PATH]; it is the one seam to migrate to a generated descriptor
 * the moment the contract is emitted.
 */
class HttpDeviceRegistrationClient(
    private val api: ApiHttpClient,
) : DeviceRegistrationClient {
    override suspend fun register(request: DeviceRegistrationRequest): Result<DeviceRegistrationResponse> =
        api
            .safeRequest<DeviceRegistrationResponse>(
                method = HttpMethodKind.POST,
                path = DEVICES_PATH,
                body = request,
            ).mapCatching { response ->
                require(response.id.isNotBlank()) { "Device registration returned no registration id" }
                response
            }

    override suspend fun unregister(registrationId: String): Result<Unit> =
        api
            .safeRequest<String>(
                method = HttpMethodKind.DELETE,
                path = "$DEVICES_PATH/$registrationId",
            ).fold(
                onSuccess = { Result.success(Unit) },
                onFailure = { error -> if (isAlreadyRemoved(error)) Result.success(Unit) else Result.failure(error) },
            )

    private fun isAlreadyRemoved(error: Throwable): Boolean = error is ApiError.Http && error.status == NOT_FOUND

    private companion object {
        /** The additive ADR-009 device-registration route (the `/api/v1` prefix is added by the client). */
        const val DEVICES_PATH = "/devices"
        const val NOT_FOUND = 404
    }
}
